feat!: enable TypeScript, provide types for all components at `@transferwise/components` (#846)
transferwise/neptune-web: packages/components/CONTRIBUTING.md

Please familiarise yourself with our [accessibility guidelines](https://github.com/transferwise/neptune-web/blob/main/ACCESSIBILITY.md) and ensure your changes meet accessibility standards.

# TypeScript Support

> We encourage you write new components on TypeScript, please consider rewrite components to TypeScript if you do major changes/refactoring.

Reviewer: I recommend being even stricter with this 😊

If you do migrate some component(s) to TypeScript please make sure to double check and if necessary remove it from [list of component for custom type generation process](https://github.com/transferwise/neptune-web/blob/main/packages/components/scripts/generate-type-declarations.js#L10).

> 

Author: link: https://github.com/transferwise/neptune-web/blob/enable-ts-for-components/packages/components/scripts/generate-type-declarations.js#L10

Please familiarise yourself with our [accessibility guidelines](https://github.com/transferwise/neptune-web/blob/main/ACCESSIBILITY.md) and ensure your changes meet accessibility standards.

# TypeScript Support

> We strongly encourage you write new components on TypeScript, please consider rewrite components to TypeScript if you do major changes/refactoring.

Reviewer 2: ```suggestion
We encourage new components to be written in TypeScript, please also consider to rewrite components in TypeScript during major refactoring.
```

Author: 👌

Author: 👌

> If you do migrate some component(s) to TypeScript please make sure to double check and if necessary remove it from [list of component for custom type generation process](https://github.com/transferwise/neptune-web/blob/main/packages/components/scripts/generate-type-declarations.js#L10).
> 

Reviewer 2: ```suggestion
While migrating a component to TypeScript please remove it, if necessary, from [list of component for custom type generation process](https://github.com/transferwise/neptune-web/blob/main/packages/components/scripts/generate-type-declarations.js#L10).

```

Author: 👌

### JS Component Rules

> If you don't rewrite component on TypeScript and just simply add new feature or fix bug make sure to ask yourself "how this would work in TypeScript world" this would help us later easier migrate this component to TS.
> 

Reviewer 2: ```suggestion
For JS new features or bug fixes please always keep in mind how they would behave in a Typescript environment as this would help us to migrate them to TS.
```

Author: 👍

> Also, please follow these rules, this will ensure that type declarations for JS components generated correctly by [`react-to-typescript-definitions`](https://www.npmjs.com/package/react-to-typescript-definitions).
> 
> Note: these rules necessary only for components that are exposed from the package, but it will great if you follow them for internal components as well for consistency and in case we decide export internal component it would be less of an effort todo that.

Reviewer 2: ```suggestion
Please follow these rules to ensure that type declarations for JS components gets generated correctly by [`react-to-typescript-definitions`](https://www.npmjs.com/package/react-to-typescript-definitions).

Note: Please try to follow these rules for both exposed and internal components to improve codebase consistency and to reduce the effort in case we decide expose an internal component.
```

Author: 👍

> - Make sure `export default` is clear, you shouldn't wrap component in any functions and other stuff, as this will lead to fully skipping generating `.d.ts` file or types with TS error `JSX element type 'SomeComponent' does not have any construct or call signatures.`
> 

Reviewer 2: ```suggestion
- Please avoid wrapping components and have a clear `export default` as otherwise this will lead to fully skipping generating `.d.ts` file or types with TS error `JSX element type 'SomeComponent' does not have any construct or call signatures.`
```

Author: 🆗

```javascript
// don't
export default injectIntl(SomeComponent);
export default withSomeHoC(SomeComponent);

// do 
export default SomeComponent;
```
- pass raw values into `PropTypes.oneOf` at `propTypes`
```javascript
// don't
SomeComponent.propTypes = {
  status: PropTypes.oneOf([Size.EXTRA_SMALL, Size.SMALL, Size.MEDIUM]),
  position: PropTypes.oneOf(Object.values(Position)),
}
SomeComponent.defaultProps = {
  size: Size.SMALL,
}

// do
SomeComponent.propTypes = {
  size: PropTypes.oneOf(['xs', 'sm', 'md']),
  position: PropTypes.oneOf(['top', 'left', 'right', 'bottom']),
}
SomeComponent.defaultProps = {
  size: Size.SMALL,
}
```

> - no need to create following props inside components, they won't work at TypeScript consumer projects

Reviewer 2: ```suggestion
- There is no need to create static properties as they won't work in TypeScript consumer apps
```

Author: 👍

```javascript
// inside Component.js

// don't
Component.Size = Size;
Component.Position = Position;
```
- define `prop-types` as `PropTypes`
```javascript
// don't
import Types from 'prop-types';

// do
import PropTypes from 'prop-types';
```
- avoid using custom validator in prop-types, this will lead to type `any` for those props
```javascript
// don't
onClick: requiredIf(PropTypes.func, ...),
content: deprecated(PropTypes.node, ...),
```

> - Ultimately, alwasy check generated types, run `yarn build` inside `packages/components` folder then head to `/build/types/index.d.ts` file, from there, navigate to your component `.d.ts` file and check accuracy of types, if types aren't what meant to be then tweak prop-types in the component and repeat these steps again.

Reviewer 2: ```suggestion
- Ultimately, always check the generated types by: 
  - run `yarn build` inside `packages/components`
  - head to `/build/types/index.d.ts`,
  - search for your component type declaration in `.d.ts` 
  - if types aren't the expected ones please tweak the prop-types in the component and repeat these steps again.
```
